### PR TITLE
Mock transport with pc frames

### PR DESCRIPTION
**Pull request by migueleliasweb**
An attempt to address https://github.com/google/go-github/issues/1800

- Uses `runtime.Frames` to get a the caller method in the SDK and use it to mock a response
- Uses code generation to automagically generate all possible matches for all exported methods currently in the SDK

### CODE REVIEW

**Comment by migueleliasweb** [comment]
I've added the code generation to all methods now. Let me know your thoughts @gmlewis :+1: .

File: mock/gen/gen-request-matches.go [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+			return !strings.HasSuffix(fi.Name(), "_test.go") &&
+				!strings.HasPrefix(fi.Name(), "gen-") &&
+				!strings.HasPrefix(fi.Name(), "github-")
+		},
+		0,
+	)
+
+	if err != nil {
+		log.Fatal(err)
+		return
```
Comment (gmlewis) [inline mock/gen/gen-request-matches.go:78]: Please delete this `return` line.

File: mock/gen/gen-request-matches.go [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+}
+
+func getServiceMethodCall(f *ast.File) <-chan RequestMatchPair {
+	returnChan := make(chan RequestMatchPair)
+
+	go func(c chan RequestMatchPair) {
+		defer close(returnChan)
+
+		for _, decl := range f.Decls {
+			if fn, ok := decl.(*ast.FuncDecl); ok {
```
Comment (gmlewis) [inline mock/gen/gen-request-matches.go:102]:
```suggestion
			fn, ok := decl.(*ast.FuncDecl)
			if !ok {
				continue
			}

```

File: mock/gen/gen-request-matches.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+				if !fn.Name.IsExported() {
+					continue
+				}
+
+				// skip empty receivers or empty lists
+				if fn.Recv == nil || len(fn.Recv.List) == 0 {
+					continue
+				}
+
+				if receiverType, ok := fn.Recv.List[0].Type.(*ast.StarExpr); ok {
```
Comment (gmlewis) [inline mock/gen/gen-request-matches.go:113]:
```suggestion
				receiverType, ok := fn.Recv.List[0].Type.(*ast.StarExpr)
				if !ok {
					continue
				}

```

File: mock/gen/gen-request-matches.go [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+					if receiverTypeIndent, ok := receiverType.X.(*ast.Ident); ok {
```
Comment (gmlewis) [inline mock/gen/gen-request-matches.go:114]:
```suggestion
					receiverTypeIndent, ok := receiverType.X.(*ast.Ident)
					if !ok || receiverTypeIndent.Name == "" {
						continue
					}

```

File: mock/gen/gen-request-matches.go [hunk 5]
```diff
@@ -0,0 +1,174 @@
...
+						if receiverTypeIndent.Name != "" {
```
Comment (gmlewis) [inline mock/gen/gen-request-matches.go:115]: This line can then be deleted if you take my suggestions above.

File: mock/gen/gen-request-matches.go [hunk 6]
```diff
@@ -0,0 +1,174 @@
...
+							varName := fmt.Sprintf(
+								"RequestMatch%s%s",
+								strings.Replace(
```
Comment (gmlewis) [inline mock/gen/gen-request-matches.go:118]: After the reduction in indentation above, this `strings.Replace` could all fit nicely on a single line.

File: mock/gen/gen-request-matches.go [hunk 7]
```diff
@@ -0,0 +1,174 @@
...
+
+func renderTemplate(
+	targetPackage string,
+	requestMatches []RequestMatchPair,
+) {
+	sort.SliceStable(requestMatches, func(i, j int) bool {
+		return requestMatches[i].Value < requestMatches[j].Value
+	})
+
+	// fmt.Println("requestMatches:", requestMatches)
```
Comment (gmlewis) [inline mock/gen/gen-request-matches.go:156]: Please remove this debugging line.

File: mock/gen/gen-request-matches.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+
+	buf := bytes.NewBufferString("")
```
Comment (gmlewis) [inline mock/gen/gen-request-matches.go:158]:
```suggestion
	var buf bytes.Buffer
```

File: mock/gen/gen-request-matches.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+
+	template.Must(template.New("source").Parse(source)).Execute(
+		buf,
```
Comment (gmlewis) [inline mock/gen/gen-request-matches.go:161]:
```suggestion
		&buf,
```

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @migueleliasweb .  I'm liking this approach.
You will also want to add a new generator to this PR here:
https://github.com/google/go-github/blob/master/github/github.go#L6-L7

**Comment by gmlewis** [comment]
After further consideration, I think that @willnorris is correct in his [comment](https://github.com/google/go-github/pull/1980#issuecomment-881619560) that this would be a great utility to place in an external helper repo.

So my recommendation is that we close this PR and also close #1800 and point to a new helper repo in this repo's README.md once it has been implemented.

**Comment by migueleliasweb** [comment]
Thanks @gmlewis for the extensive review!

I left a comment in the initial PR (https://github.com/google/go-github/pull/1980) so you and Will can talk more. I will wait until the decision is made to move this to a new repo or not before proceeding with the rest of the implementation.

**Pull request closed without merging**